### PR TITLE
#321 add rules section to Kernel Transformations

### PR DESCRIPTION
**Pull request by arporter**
Documents the rules we will apply to kernels that are candidates for transformation.

### CODE REVIEW

**Comment by arporter** [comment]
Either @sergisiso or @rupertford could review I think.

**Comment by arporter** [comment]
Also, please could you take a look @TeranIvy to check on the implications for LFRic?

**Comment by TeranIvy** [comment]
The rules seem sensible to me. However, I am not sure about availability of the UM modules code that LFRic uses (possible logistic and legal issues) so I asked @stevemullerworth for his input.

**Comment by rupertford** [comment]
Back to @arporter (and @sergisiso if he wants to comment)

File: doc/transformations.rst [hunk 1]
```diff
@@ -180,6 +180,50 @@ transformed version of that kernel exists and does not match that
...
+
+Kernels that are to be transformed are subject to additional
+restrictions. These rules are intended to make kernel transformations
+as robust as possible, in particular by limiting the amount of
+code that must be parsed by PSyclone (via fparser). The rules are
+as follows:
+
+1) Any variable accessed by a kernel must either be explicitly declared
+   or named in the ``only`` clause of a module ``use`` statement within
+   the scope of the subroutine containing the kernel implementation;
```
[the remark this answers is not in the document]
Comment (sergisiso) [reply doc/transformations.rst]: "Any variable" to "Any variable or procedure" (as I understand the same restriction applies for both?)

File: doc/transformations.rst [hunk 2]
```diff
@@ -180,6 +180,50 @@ transformed version of that kernel exists and does not match that
...
+
+2) The full Fortran source of a kernel must be available to PSyclone.
+   This includes the source of any modules from which it accesses
+   either routines or data.
```
Comment (sergisiso) [inline doc/transformations.rst]: I am not convinced with "The full Fortran source" as in the case of routines there are several Fortran mechanisms to soften the constraint: interfaces, public keyword (?), c_bindings.

An alternative would be to link 2) to the previous point and ask for the declarations of the named instances in the `use` statements to be available to (or discoverable by) PSyclone.

**Comment by stevemullerworth** [comment]
> The rules seem sensible to me. However, I am not sure about availability of the UM modules code that LFRic uses (possible logistic and legal issues) so I asked @stevemullerworth for his input.

If I understand the question correctly, I think the answer is that we would never expect to run PSyclone on a file if we have not also extracted and made available all the other code the file is dependent on. Therefore there would be no PSyclone issue other than knowing where to look for the code.

File: doc/transformations.rst [hunk 3]
```diff
@@ -180,6 +180,50 @@ transformed version of that kernel exists and does not match that
 created by the current transformation then PSyclone will raise an
 exception.
 
+Rules
++++++
+
+Kernels that are to be transformed are subject to additional
```
[the remark this answers is not in the document]
Comment (arporter) [reply doc/transformations.rst]: Done.

File: doc/transformations.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (arporter) [reply doc/transformations.rst]: @rupertford, for the first two questions I refer you to my text :-) Good point about common and extern - I shall say that we don't support them. I shall generalise to mention procedures and the fact that we support intrinsics (which I think we should).

File: doc/transformations.rst (hunk 2 above)
Comment (arporter) [reply doc/transformations.rst]: Thanks @sergisiso. I've had a think about this and the problem is that we need to be sure that, when transforming kernels for either OpenACC or OpenCL, we've captured all use-dependencies in the call-tree of a kernel. However, having written that I can see that this is only a condition for those two cases - if we were just manipulating the form of a kernel routine then perhaps just having the interface information available would be sufficient. Given that OpenACC and OpenCL are our immediate targets, I think keeping this restriction as it is is OK. We can always relax it in future if/when we are doing other sorts of kernel transformation and it becomes a problem.

**Comment by arporter** [comment]
Thanks for all the comments. I've updated the text to be more explicit and also to forbid COMMON and EXTERN. Ready for another review.

**Pull request merged**